verifier/verifier.go

```diff
@@ -102,56 +128,111 @@ func Verify(layout *Layout, attestations map[string]*dsse.Envelope, parameters m
 			return fmt.Errorf("no claims found for step %s", step.Name)
 		}
 
-		for _, expectedPredicate := range step.ExpectedPredicates {
-			if expectedPredicate.Threshold == 0 {
-				expectedPredicate.Threshold = 1
+		if step.Threshold == 0 {
+			step.Threshold = 1
+		}
+
+		trustedStatements := getPredicates(stepStatements, step.Functionaries)
+		if len(trustedStatements) < step.Threshold {
+			return fmt.Errorf("threshold not met for step %s", step.Name)
+		}
+
+		// TODO: reduce statements if they're identical to avoid checking all of
+		// them
+		// See in-toto 1.0
+
+		acceptedPredicates := 0
+		failedChecks := []error{}
+		for functionary, statement := range trustedStatements {
+			log.Infof("Verifying claim for step '%s' of type '%s' by '%s'...", step.Name, step.ExpectedPredicateType, functionary)
+			failed := false
+
+			// Check the predicate type matches the expected value in the layout
+			if step.ExpectedPredicateType != statement.PredicateType {
+				failed = true
+				failedChecks = append(failedChecks, fmt.Errorf("for step %s, statement with unexpected predicate type %s found", step.Name, statement.PredicateType))
 			}
 
-			matchedPredicates := getPredicates(stepStatements, expectedPredicate.PredicateType, expectedPredicate.Functionaries)
-			if len(matchedPredicates) < expectedPredicate.Threshold {
-				return fmt.Errorf("threshold not met for step %s", step.Name)
+			// Check materials and products
+			if err := applyArtifactRules(statement, step.ExpectedMaterials, step.ExpectedProducts, claims); err != nil {
+				failed = true
+				failedChecks = append(failedChecks, fmt.Errorf("for step %s, claim by %s failed artifact rules: %w", step.Name, functionary, err))
 			}
 
-			failedChecks := []error{}
-			acceptedPredicates := 0
-			for functionary, statement := range matchedPredicates {
-				log.Infof("Verifying claim for step '%s' of type '%s' by '%s'...", step.Name, expectedPredicate.PredicateType, functionary)
-				failed := false
+			input, err := getActivation(statement)
+			if err != nil {
+				return err
+			}
 
-				if err := applyArtifactRules(statement, step.ExpectedMaterials, step.ExpectedProducts, claims); err != nil {
-					failed = true
-					failedChecks = append(failedChecks, fmt.Errorf("for step %s, claim by %s failed artifact rules: %w", step.Name, functionary, err))
-				}
+			// Check attribute rules
+			if err := applyAttributeRules(env, input, step.ExpectedAttributes); err != nil {
+				failed = true
+				failedChecks = append(failedChecks, fmt.Errorf("for step %s, claim by %s failed attribute rules: %w", step.Name, functionary, err))
+			}
 
-				input, err := getActivation(statement)
+			// Examine collector claims in attestation collection
+			if step.ExpectedPredicateType == witnessattestation.CollectionType {
```

Comment:
I think this special casing is really what I'm referring to when I say that we ought to be generalizing. In some sense, this code is really doing the sort of pre-processing or inspection that I think ultimately should be implemented as a separate module. We might still provide this module for Witness as part of the attestation-verifier package, but this is where I think we can start figuring out what the common APIs for inspections 2.0 .